src/JuliusSweetland.OptiKey/Models/XmlKeyboard.cs

```diff
@@ -0,0 +1,246 @@
+
+        public string Name
+        { get; set; }
+
+        public string Symbol
+        { get; set; }
+
+        public double? SymbolMargin
+        { get; set; }
+
+        protected static readonly ILog Log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+
+        [XmlIgnore]
+        public Thickness? BorderThickness
+        { get; set; }
+
+        [XmlElement("BorderThickness")]
+        public string BorderThicknessAsString
+        {
+            get { return BorderThickness.ToString(); }
+            set {
+                try
+                {
+                    ThicknessConverter thicknessConverter = new ThicknessConverter();
+                    BorderThickness = (Thickness)thicknessConverter.ConvertFromString(value);
+                }
+                catch (System.FormatException)
+                {
+                    Log.ErrorFormat("Cannot interpret \"{0}\" as thickness", value);                
+                }
+            }
+        }        
+        
+        [XmlIgnore]
+        public bool Hidden
+        { get; set; }
+
+        [XmlElement("HideFromKeyboardMenu")]
+        public string HiddenBoolAsString
+        {
+            get { return this.Hidden ? "True" : "False"; }
+            set { this.Hidden = XmlUtils.ConvertToBoolean(value); }
+        }        
+        
+        public static XmlKeyboard ReadFromFile(string inputFilename)
+        {
+            XmlKeyboard keyboard;
+
+            // If no extension given, try ".xml"
+            string ext = Path.GetExtension(inputFilename);
+            bool exists = File.Exists(inputFilename);
+            if (!File.Exists(inputFilename) &&
+                String.IsNullOrEmpty(Path.GetExtension(inputFilename)))
+            {
+                inputFilename += ".xml";
+            } 
+
+            // Read in XML file (may throw)
+            XmlSerializer serializer = new XmlSerializer(typeof(XmlKeyboard));
+            using (FileStream readStream = new FileStream(@inputFilename, FileMode.Open))
+            {
+                keyboard = (XmlKeyboard)serializer.Deserialize(readStream);
+            }
+
+            return keyboard;
+        }
+    }
+
+    public class XmlGrid
+    {
+        public int Rows
+        { get; set; }
+
+        public int Cols
+        { get; set; }
+
+    }
+
+    public class XmlKeys
+    {
+        [XmlElement(ElementName = "BehaviourKey")]
+        public List<XmlBehaviourKey> BehaviourKeys
+        { get; set; }
+
+        [XmlElement(ElementName = "TextKey")]
+        public List<XmlTextKey> TextKeys
+        { get; set; }
+
+        [XmlElement(ElementName = "LinkKey")]
+        public List<XmlLinkKey> LinkKeys
+        { get; set; }
+
+        [XmlIgnore]
+        public int Count
+        {
+            get
+            {
+                return BehaviourKeys.Count + TextKeys.Count + LinkKeys.Count;
+            }
+        }
+    }
+
+    public class XmlLinkKey : XmlKey
```

Comment:
Rename to XmlChangeKeyboardKey?
Also split out into separate files per class (bundle into a new folder?)